### PR TITLE
Core functionality of bionic slots system (again)

### PR DESCRIPTION
**Pull request by illi-kun**
3rd attempt to implement bionic slots system, this time without contradictory UI changes.

Occupied bodyparts are listed in:
- item description menu:
  item
- bionic description (versions for small & comfortable terminal sizes): 
  list_bionic_min
  list_bionic_max

Color coding of CBMs in inventory is improved, red color of CBM unit in inventory means "some bodyparts have not enough space for installation":
inv

Attempting to install "red" bionic shows actual issues / lack of bodypart slots:
comply

New tab for CBMs review & management and further upgrades of existing bionic menu will be implemented separately (if it's acceptable).

### CODE REVIEW

**Comment by illi-kun** [comment]
Previous discussion about these changes is available here: #15937.
Feature request similar to what is implemented here is #11781.

File: src/bionics.cpp [hunk 1]
```diff
@@ -459,7 +463,11 @@ void player::power_bionics()
                                         power_only_desc.str() );
             }
             ypos += fold_and_print( w_description, ypos, 0, DESCRIPTION_WIDTH, c_ltblue,
-                                    bionics[( *current_bionic_list )[cursor]->id].description );
+                                    bionics[( *current_bionic_list )[cursor]->id].description ) + 1;
+            ypos += fold_and_print( w_description, ypos, 0, DESCRIPTION_WIDTH, c_ltgray,
+                                    list_occupied_bps( ( *current_bionic_list )[cursor]->id,
+                                    _( "This bionic occupies the following body parts:" ),
+                                    ( getmaxy( w_description ) > ypos + ( int )num_bp + 1 ) ).c_str() );
```
Comment (codemime) [inline src/bionics.cpp]: It's quite hard to comprehend the meaning without looking at the declaration of `list_occupied_bps()`
A boolean flag can make the task much easier:
```c++
const bool one_per_line = ypos + ( int )num_bp + 1 < getmaxy( w_description ); 
...
list_occupied_bps( ..., one_per_line ).c_str() 
```

File: src/bionics.cpp [hunk 2]
```diff
@@ -1532,9 +1521,22 @@ bool player::install_bionics( const itype &type, int skill_level )
                                               difficult );
     }
 
-    if( !query_yn(
-            _( "WARNING: %i percent chance of genetic damage, blood loss, or damage to existing bionics! Install anyway?" ),
-            100 - chance_of_success ) ) {
+    std::map<body_part, int> issues = bionic_installation_issues( bioid );
```
Comment (codemime) [inline src/bionics.cpp]: Can become:
```c++
const auto &issues = bionic_installation_issues( bioid );
```
I suppose the ampersand can be omitted, though (due to compiler's optimisation).

File: src/bionics.cpp [hunk 3]
```diff
@@ -1532,9 +1521,22 @@ bool player::install_bionics( const itype &type, int skill_level )
...
+    // show all requirements which are not satisfied
+    if( !issues.empty() ) {
+        std::string detailed_info;
+        for( auto& elem : issues ) {
+            //~ <Body part name>: <number of slots> more slot(s) needed.
+            detailed_info += string_format( _( "\n%s: %i more slot(s) needed." ),
+                                            body_part_name_as_heading( elem.first, 1 ).c_str(),
+                                            elem.second );
+        }
```
Comment (codemime) [inline src/bionics.cpp]: You can probably reuse `list_occupied_bps()` (the code looks pretty similar) and write something like this:
```c++
popup( list_bionic_bps( issues, ... ) ... );
```

File: src/bionics.cpp [hunk 4]
```diff
@@ -1705,6 +1707,92 @@ void bionics_install_failure( player *u, int difficulty, int success )
     }
 }
 
+std::string list_occupied_bps( std::string bio_id, std::string intro, bool one_per_line )
```
Comment (codemime) [inline src/bionics.cpp]: Please, pass strings by reference: 
```c++
( const std::string &bio_id, ...
```

File: src/bionics.cpp [hunk 5]
```diff
@@ -1705,6 +1707,92 @@ void bionics_install_failure( player *u, int difficulty, int success )
...
+
+int player::get_used_bionics_slots( const body_part bp ) const
+{
+    int used_slots = 0;
+    for( auto &bio : my_bionics ) {
+        auto search = bionics[bio.id].occupied_bodyparts.find( bp );
+        if( search != bionics[bio.id].occupied_bodyparts.end() ) {
+            used_slots += search->second;
+        }
+     }
```
Comment (codemime) [inline src/bionics.cpp]: Extra indent?

File: src/bionics.cpp [hunk 6]
```diff
@@ -1705,6 +1707,92 @@ void bionics_install_failure( player *u, int difficulty, int success )
...
+}
+
+std::map<body_part, int> player::bionic_installation_issues( std::string bioid )
+{
+    std::map<body_part, int> issues;
+    for( auto& elem : bionics[ bioid ].occupied_bodyparts ) {
+        if( !has_enough_slots( elem.first, elem.second ) ) {
+            issues.emplace( elem.first, (int)elem.second -
+                            get_free_bionics_slots( elem.first ) );
+        }
```
Comment (codemime) [inline src/bionics.cpp]: I'd rewrite the inner snippet like this:
```c++
const int lacked_slots = elem.second - get_free_bionics_slots( elem.first );
if( lacked_slots > 0 ) {
    issues.emplace( elem.first, lacked_slots );
}
```
It's slightly more expressive, saves you one redundant call to `get_free_bionics_slots()`, and enables getting rid of `has_enough_slots()`.

File: src/bionics.cpp [hunk 7]
```diff
@@ -1705,6 +1707,92 @@ void bionics_install_failure( player *u, int difficulty, int success )
...
+    case bp_leg_l:
+    case bp_leg_r:
+        return 30;
+
+    case bp_foot_l:
+    case bp_foot_r:
+        return 7;
+
+    case num_bp:
+        return INT_MAX;
```
Comment (codemime) [inline src/bionics.cpp]: Please, consider `debugmsg()` here.

File: src/bionics.cpp [hunk 8]
```diff
@@ -1850,14 +1938,28 @@ void load_bionic( JsonObject &jsobj )
     bool faulty = jsobj.get_bool( "faulty", false );
     bool power_source = jsobj.get_bool( "power_source", false );
 
+    JsonArray jsarr;
+    std::map<body_part, size_t> occupied_bodyparts;
+    jsarr = jsobj.get_array( "occupied_bodyparts" );
```
Comment (codemime) [inline src/bionics.cpp]: Can become:
```c++
JsonArray jsarr = jsobj.get_array( "occupied_bodyparts" );
```

File: src/bionics.h [hunk 9]
```diff
@@ -5,13 +5,10 @@
...
 #include <string>
 
-/* Thought: Perhaps a HUD bionic that changes the display of the game?
- * Showing more information or something. */
-
 struct bionic_data {
     bionic_data() = default;
     bionic_data( std::string nname, bool ps, bool tog, int pac, int pad, int pot,
-                 int ct, int cap, std::string desc, bool fault );
+                 int ct, int cap, std::string desc, bool fault, std::map<body_part, size_t> bps );
```
Comment (codemime) [inline src/bionics.h]:
```c++
..., const std::map<body_part, size_t> &bps );
```

File: src/bionics.cpp [hunk 10]
```diff
@@ -91,17 +92,20 @@ bionic_data const &bionic_info( std::string const &id )
...
+        "bad bionic", false, false, 0, 0, 0, 0, 0, "bad_bionic", false, {{bp_torso, 0}}
+    };
     return null_value;
 }
 
 void bionics_install_failure( player *u, int difficulty, int success );
 
 bionic_data::bionic_data(std::string nname, bool ps, bool tog, int pac, int pad, int pot,
-                         int ct, int cap, std::string desc, bool fault
+                         int ct, int cap, std::string desc, bool fault, std::map<body_part, size_t> bps
```
Comment (codemime) [inline src/bionics.cpp]: ..., const std::map<body_part, size_t> &bps );

File: src/bionics.cpp [hunk 11]
```diff
@@ -1850,14 +1938,28 @@ void load_bionic( JsonObject &jsobj )
...
+    std::map<body_part, size_t> occupied_bodyparts;
+    jsarr = jsobj.get_array( "occupied_bodyparts" );
+    if( !jsarr.empty() ) {
+        while( jsarr.has_more() ) {
+            JsonArray ja = jsarr.next_array();
+            occupied_bodyparts.emplace( get_body_part_token( ja.get_string( 0 ) ),
+                                        ja.get_int( 1 ) );
+        }
+    } else {
+        occupied_bodyparts.emplace( bp_torso, 0 );
```
Comment (codemime) [inline src/bionics.cpp]: Should probably be reported via `debugmsg()`.

Comment (illi-kun) [inline src/bionics.cpp]: I disagree. Some bionics have no occupied body parts defined and that's fine (I'm talking about reactor upgrade and power storage units).

File: src/bionics.cpp (hunk 5 above)
Comment (illi-kun) [inline src/bionics.cpp]: Wow, good catch!

File: src/bionics.cpp (hunk 3 above)
Comment (illi-kun) [inline src/bionics.cpp]: Here we're listing only bodyparts which have a lack of free slots.

```list_occupied_bps``` obtains ```bionic id``` as an input and generates full list of occupied parts by this bionic. Your suggestion requires expansion/modification of ```list_occupied_bps``` function, so I'd prefer to keep it as is.

File: src/bionics.cpp (hunk 11 above)
Comment (codemime) [inline src/bionics.cpp]: Only three out of 109, so it's not a reasonable default. You could either define `["TORSO", 0]` explicitly in JSON and still report errors (more reliable: people won't forget about the slot system while introducing/changing bionics), or simply leave the map empty (no overhead). Both options are IMHO better.

File: src/bionics.cpp (hunk 3 above)
Comment (codemime) [inline src/bionics.cpp]: Both the function and the snippet do essentially the same thing, but have different implementation which can be easily unified. Though it's a minor thing.

File: src/bionics.cpp (hunk 11 above)
Comment (illi-kun) [inline src/bionics.cpp]: Okay, I think you're right. BTW, current code and your first solution causes this inappropriate note for CBMs listed above:
bad_bp

For avoiding this, I'll implement your latter solution.

Thanks.

**Comment by Coolthulhu** [comment]
I can't find any info on how many slots I have left, except by trying to install a CBM that goes over the limit. This is mandatory functionality before slot limiting can be enabled in non-debug mode.

To simplify the development, you can add a DEBUG_CBM_SLOTS mutation and check for it. At first, it would enforce the slot limit (with non-debug characters not respecting it), later it could be reversed (ie. characters with said mutation would not respect it) and then it would be useful for debugging huge sets of bionics at once.

**Comment by illi-kun** [comment]
Sounds reasonable.

**Comment by illi-kun** [comment]
@Coolthulhu done.

**Pull request merged**